### PR TITLE
[Qt] honor current network when creating autostart link

### LINKED ISSUE #5778: autostart shortcut should honor -testnet
When checking "Start Bitcoin on system login" the shortcut created doesn't include -testnet. This means that the shortcut created by the bitcoin-core testnet program starts a mainnet client.

### PR DESCRIPTION
**Pull request by Diapolo**
- creates a "Bitcoin (testnet).lnk" when on testnet or a "Bitcoin
  (regtest).lnk, when on regtest
- fixes #5778

### CODE REVIEW

**Comment by Diapolo** [comment]
Hopefully fixed the Travis build and also added the change for Linux (needs testing).

File: src/qt/guiutil.cpp [hunk 1]
```diff
@@ -597,20 +600,33 @@ bool SetStartOnSystemStartup(bool fAutoStart)
             TCHAR pszExePath[MAX_PATH];
             GetModuleFileName(NULL, pszExePath, sizeof(pszExePath));
 
-            TCHAR pszArgs[5] = TEXT("-min");
+            // Start client minimized
+            QString strArgs = "-min";
+            // Start testnet client, if we are currently running testnet
+            if (Params().NetworkIDString() == "test")
```
Comment (laanwj) [inline src/qt/guiutil.cpp]: You could generalize this by passing the current value of testnet, or even regtest to the link, 
```c++
strprintf("-testnet=%d -regtest=%d", GetBoolArg("-testnet"), GetBoolArg("-regtest"))
```
Not that I think there is much use in an autostarting regtest instance.

**Comment by laanwj** [comment]
Failing travis on windows
```
 CXX qt/qt_libbitcoinqt_a-guiutil.o
qt/guiutil.cpp: In function ‘bool GUIUtil::SetStartOnSystemStartup(bool)’:
qt/guiutil.cpp:610:61: error: cannot convert ‘WCHAR* {aka wchar_t*}’ to ‘TCHAR* {aka char*}’ in initialization
qt/guiutil.cpp:611:42: error: no matching function for call to ‘QString::toWCharArray(TCHAR*&)’
qt/guiutil.cpp:611:42: note: candidate is:
/home/travis/build/bitcoin/bitcoin/depends/x86_64-w64-mingw32/include/QtCore/qstring.h:824:12: note: int QString::toWCharArray(wchar_t*) const
/home/travis/build/bitcoin/bitcoin/depends/x86_64-w64-mingw32/include/QtCore/qstring.h:824:12: note: no known conversion for argument 1 from ‘TCHAR* {aka char*}’ to ‘wchar_t*’
make[2]: *** [qt/qt_libbitcoinqt_a-guiutil.o] Error 1
make[2]: Leaving directory `/home/travis/build/bitcoin/bitcoin/bitcoin-x86_64-w64-mingw32/src'
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory `/home/travis/build/bitcoin/bitcoin/bitcoin-x86_64-w64-mingw32/src'
make: *** [check-recursive] Error 1
```

**Comment by Diapolo** [comment]
Yeah I already saw this... will need to rework :).

**Comment by Diapolo** [comment]
Who is able to test this :)?

File: src/qt/guiutil.cpp (hunk 1 above)
Comment (laanwj) [inline src/qt/guiutil.cpp]: If you do keep it like this based on special-casing the network name, make sure that you avoid doing anything if the net is not `main` or `test`

File: src/qt/guiutil.cpp [hunk 2]
```diff
@@ -597,20 +600,40 @@ bool SetStartOnSystemStartup(bool fAutoStart)
...
+                strArgs += " -testnet";
+
+#ifdef UNICODE
+            // Convert the QString to TCHAR*
+            TCHAR* args = new TCHAR[strArgs.length() + 1];
```
Comment (laanwj) [inline src/qt/guiutil.cpp]: We should use RAII here, could be done in various ways. Most obvious is boost::scoped_array, e.g.

    scoped_array<TCHAR> args (new TCHAR[strArgs.length() + 1]);

Another option would be going through std::wstring, e.g.

    std::wstring wstrArgs = strArgs.toStdWString();
    TCHAR *args = W2T(wstrArgs.c_str());

In this case the manual appending of a NULL byte can be avoided, and the code can be shorter.

**Comment by Diapolo** [comment]
@laanwj I choose the scoped_array, because the other macro wasn't available here. I also updated to your generic approach for testnet/tegtest.

**Comment by laanwj** [comment]
utACK

**Comment by laanwj** [comment]
@tykling  can you test this?

**Comment by tykling** [comment]
I have no means of building this, sorry, but from reviewing the code it appears to do exactly what is should :) Is there somewhere I can download the executable travis built and try that?

**Comment by jonasschnelli** [comment]
Tested ACK.
Tested ´regtest´ and ´testnet´ (parallel).
Checkbox also reacts correct if testnet/regtest link is already there.

Binaries to test: https://builds.jonasschnelli.ch/pulls/5793/

bildschirmfoto 2015-03-05 um 07 01 39

After restart:
bildschirmfoto 2015-03-05 um 07 07 52

**Pull request merged**